### PR TITLE
data_manager for CAT

### PR DESCRIPTION
**Pull request by jj-umn**
Either downloads prebuilt CAT reference data from: https://tbb.bio.uu.nl/bastiaan/CAT_prepare/
or build a new reference from NCBI nr and taxonomy using command:
CAT prepare --fresh

FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: data_managers/data_manager_cat/data_manager/data_manager_cat.xml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<tool id="data_manager_cat" name="CAT DB" version="@VERSION@.0" tool_type="manage_data">
+    <description>Install a new CAT database</description>
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements" />
+    <expand macro="version_command" />
+    <command detect_errors="exit_code"><![CDATA[
+#import json, os
+#set params = json.loads(open(str($out_file)).read())
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager/data_manager_cat.xml]: I think you can skip this part and just do `python '${__tool_directory__}/data_manager_cat.py' --config_file '$out_file' --install_path '$out_file.extra_files_path '`

File: data_managers/data_manager_cat/data_manager/data_manager_cat.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    dm_dict['data_tables'] = dm_dict.get('data_tables', {})
+    data_table = 'cat_database'
+    dm_dict['data_tables'][data_table]\
+        = dm_dict['data_tables'].get(data_table, [])
+    data_table_entry = dict(value=cat_dir, name=cat_dir,
+                            database_folder=os.path.join(cat_dir, cat_db),
+                            taxonomy_folder=os.path.join(cat_dir, tax_db))
+    dm_dict['data_tables'][data_table].append(data_table_entry)
+    # save info to json file
+    open(args.config_file, 'wb').write(json.dumps(dm_dict))
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager/data_manager_cat.py]:
```suggestion
    open(args.config_file, 'w').write(json.dumps(dm_dict, sort_keys=True))
```

File: data_managers/data_manager_cat/data_manager/data_manager_cat.xml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    ]]></command>
+    <inputs>
+        <conditional name="db">
+            <param name="src" type="select" label="Download or Build DBs">
+                <option value="download">download</option>
+                <option value="build">build</option>
+            </param>
+            <when value="download">
+                <param name="db_url" type="text" label="DB rul at https://tbb.bio.uu.nl/bastiaan/CAT_prepare/"
+                     help="example: https://tbb.bio.uu.nl/bastiaan/CAT_prepare/CAT_prepare_20190719.tar.gz">
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager/data_manager_cat.xml]: Could you build a tiny test database for a test ?

File: data_managers/data_manager_cat/data_manager/macros.xml [hunk 4]
```diff
@@ -0,0 +1,343 @@
+<macros>
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager/macros.xml]: I assume there is another tools that uses these macros ?

File: data_managers/data_manager_cat/data_manager/data_manager_cat.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def cat_prepare(install_dir):
+    cmd = ['CAT', 'prepare' '--fresh', '-q']
+    cmd_stdout = tempfile.NamedTemporaryFile()
+    cmd_stderr = tempfile.NamedTemporaryFile()
+    return_code = subprocess.call(cmd, shell=True, cwd=install_dir,
+                                  stdout=cmd_stdout, stderr=cmd_stderr)
+    if return_code:
+        msg = "stdout:\n%s\nstderr:\n%s" % (cmd_stdout.read(),
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager/data_manager_cat.py]:
```suggestion
        msg = "stdout:\n%s\nstderr:\n%s" % (cmd_stdout.read().decode('utf-8'),
```

File: data_managers/data_manager_cat/data_manager/data_manager_cat.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                                            cmd_stderr.read())
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager/data_manager_cat.py]:
```suggestion
                                            cmd_stderr.read().decode('utf-8'))
```

**Comment by jj-umn** [comment]
@mvdbeek I made the changes you suggested and added a test case.  Thanks.

File: data_managers/data_manager_cat/data_manager_conf.xml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0"?>
+<data_managers>
+  <data_manager tool_file="data_manager/data_manager_cat.xml" id="data_manager_cat" >
+    <data_table name="cat_database">  <!-- Defines a Data Table to be modified. -->
+      <output> <!-- Handle the output of the Data Manager Tool -->
+        <column name="value" /> <!-- columns that are going to be specified by the Data Manager Tool -->
+        <column name="name" />  <!-- columns that are going to be specified by the Data Manager Tool -->
+        <column name="database_folder" output_ref="out_file" >
+          <move type="directory" relativize_symlinks="True">
+            <src >${database_folder}</target>
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager_conf.xml]: I think there's a tag mismatch here that leads to the tests not getting executed:
```
2019-12-04 16:19:00,805 ERROR [galaxy.util] Error parsing file /home/travis/build/galaxyproject/tools-iuc/data_managers/data_manager_cat/data_manager/../data_manager_conf.xml
Traceback (most recent call last):
  File "/tmp/tmpKRSrHI/galaxy-dev/lib/galaxy/util/__init__.py", line 236, in parse_xml
    root = tree.parse(fname, parser=ElementTree.XMLParser(target=DoctypeSafeCallbackTarget()))
  File "/opt/python/2.7.15/lib/python2.7/xml/etree/ElementTree.py", line 656, in parse
    parser.feed(data)
  File "/opt/python/2.7.15/lib/python2.7/xml/etree/ElementTree.py", line 1659, in feed
    self._raiseerror(v)
  File "/opt/python/2.7.15/lib/python2.7/xml/etree/ElementTree.py", line 1523, in _raiseerror
    raise err
ParseError: mismatched tag: line 10, column 38
```
```suggestion
            <src>${database_folder}</src>
```

File: data_managers/data_manager_cat/data_manager_conf.xml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+            <target base="${GALAXY_DATA_MANAGER_DATA_PATH}">CAT/${database_folder}</target>
+          </move>
+          <value_translation>${GALAXY_DATA_MANAGER_DATA_PATH}/CAT/${database_folder}</value_translation>
+          <value_translation type="function">abspath</value_translation>
+        </column>
+        <column name="taxonomy_folder" output_ref="out_file" >
+          <move type="directory" relativize_symlinks="True">
+            <src >${taxonomy_folder}</target>
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager_conf.xml]:
```suggestion
            <src>${taxonomy_folder}</src>
```

File: data_managers/data_manager_cat/data_manager/data_manager_cat.xml [hunk 9]
```diff
@@ -6,8 +6,14 @@
     <version_command>CAT --version</version_command>
     <command detect_errors="exit_code"><![CDATA[
 mkdir -p '$out_file.extra_files_path' &&
-python '${__tool_directory__}/data_manager_cat.py' --config_file '$out_file' --install_path '$out_file.extra_files_path' 
-#if $db.src == 'download'
+python '${__tool_directory__}/data_manager_cat.py'
+ --config_file '$out_file'
+ --install_path '$out_file.extra_files_path' 
+#if $test == 'download'
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager/data_manager_cat.xml]: Can you drop this workaround, and then we download from https://github.com/galaxyproject/tools-iuc/blob/ce82f787f1f035debfc86de09a271a9ec8d27e0e/data_managers/data_manager_cat/test-data/CAT_prepare_test.tar.gz?raw=true ?

File: data_managers/data_manager_cat/data_manager/data_manager_cat.xml [hunk 10]
```diff
@@ -0,0 +1,57 @@
+<tool id="data_manager_cat" name="CAT DB" version="5.0.3.0" tool_type="manage_data">
+    <description>Install a new CAT database</description>
+    <requirements>
+        <requirement type="package" version="@VERSION@">cat</requirement>
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager/data_manager_cat.xml]:
```suggestion
        <requirement type="package" version="@VERSION@">cat</requirement>
        <requirement type="package" version="3.7">python</requirement>
```

File: data_managers/data_manager_cat/data_manager/data_manager_cat.xml [hunk 11]
```diff
@@ -0,0 +1,57 @@
+<tool id="data_manager_cat" name="CAT DB" version="5.0.3.0" tool_type="manage_data">
```
Comment (mvdbeek) [inline data_managers/data_manager_cat/data_manager/data_manager_cat.xml]:
```suggestion
<tool id="data_manager_cat" name="CAT DB" version="5.0.3.0" tool_type="manage_data" profile="18.09">
```

**Pull request merged**